src/Nervos/Transaction.cpp

```diff
@@ -0,0 +1,362 @@
+                                                          const Address& changeAddress,
+                                                          uint64_t byteFee) {
+    uint64_t availableNativeAmount = 0;
+    uint256_t availableSudtAmount = 0;
+    Cells sorted;
+    std::copy_if(signingInputCells.begin(), signingInputCells.end(), std::back_inserter(sorted),
+                 [&](const Cell& cell) { return cell.type == outputs[0].type; });
+    std::sort(sorted.begin(), sorted.end(), [&](const Cell& lhs, const Cell& rhs) {
+        return Serialization::decodeUint256(lhs.data) < Serialization::decodeUint256(rhs.data);
+    });
+    bool gotEnough = false;
+    for (auto& cell : sorted) {
+        availableSudtAmount += Serialization::decodeUint256(cell.data);
+        availableNativeAmount += cell.capacity;
+        inputCells.emplace_back(std::move(cell));
+        if (amount == 0) {
+            // Transfer maximum available tokens
+            gotEnough = true;
+        } else if (availableSudtAmount >= amount) {
+            // Transfer exact number of tokens
+            gotEnough = true;
+            uint256_t changeValue = availableSudtAmount - amount;
+            if (changeValue > 0) {
+                outputs.emplace_back(minCellCapacityForSUDT, Script(changeAddress),
+                                     Script(outputs[0].type));
+                outputsData.emplace_back(Serialization::encodeUint256(changeValue, 16));
+            }
+            break;
+        }
+    }
+    if (!gotEnough) {
+        return Common::Proto::Error_not_enough_utxos;
+    }
+    outputsData[0] = Serialization::encodeUint256(amount == 0 ? availableSudtAmount : amount, 16);
+    uint64_t fee = (sizeWithoutInputs() + inputCells.size() * sizeOfSingleInputAndWitness() +
+                    sizeOfSingleOutput(changeAddress)) *
+                   byteFee;
+    uint64_t requiredNativeAmount =
+        minCellCapacityForSUDT * outputs.size() + minCellCapacityForNativeToken + fee;
+    if (availableNativeAmount >= requiredNativeAmount) {
+        outputs.emplace_back(availableNativeAmount - requiredNativeAmount +
+                                 minCellCapacityForNativeToken,
+                             Script(changeAddress), Script());
+        outputsData.emplace_back();
+        return Common::Proto::OK;
+    } else {
+        Cells remainingCells;
+        std::copy_if(signingInputCells.begin(), signingInputCells.end(),
+                     std::back_inserter(remainingCells), [&](const Cell& cell1) {
+                         return !std::any_of(
+                             inputCells.begin(), inputCells.end(),
+                             [&](const Cell& cell2) { return cell1.outPoint == cell2.outPoint; });
+                     });
+        return planNativeTokenTransfer(remainingCells, changeAddress, byteFee);
+    }
+}
+
+Common::Proto::SigningError Transaction::sign(const std::vector<PrivateKey>& privateKeys) {
```

Comment:
same remark as above, try to separate the code logic to avoid nested loop and facilitate the review of the code

Reply:
Done